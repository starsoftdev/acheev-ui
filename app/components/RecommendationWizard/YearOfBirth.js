// @flow
import React from 'react';

import BaseModal from 'components/BaseModal';
import Button from 'components/Button';
import Input from 'components/Input';

import ALLOWED_RANGE_AGES from 'enum/allowed_range_ages';

type Props = {
  onSelect: (value: string) => void,
  // eslint-disable-next-line  react/no-unused-prop-types
  value: string | Array<string>,
};

type State = {
  value: string,
  showMessage: boolean,
  isModalOpen: boolean,
};

export default class YearOfBirth extends React.Component<Props, State> {
  state = {
    value: '',
    showMessage: false,
    isModalOpen: false,
  };

  onOpenModal = (e: SyntheticEvent<HTMLAnchorElement>) => {
    e.preventDefault();
    this.setState({ isModalOpen: true });
  };

  onCloseModal = () => {
    this.setState({ isModalOpen: false });
  };

  handleClick = () => {
    const currentYear = new Date().getFullYear();
    const max = currentYear - ALLOWED_RANGE_AGES.MIN_USER_AGE;
    const min = currentYear - ALLOWED_RANGE_AGES.MAX_USER_AGE;
    const { value } = this.state;

    const isValid = +value > min && +value < max;

    if (isValid) {
      this.props.onSelect(value);
    } else {
      this.setState({ showMessage: true });
    }
  };

  handleChange = (value: string) => {
    this.setState({ value });
  };

  render() {
    return (
      <section className="recommendationWizard__body">
        <div className="container">
          <div className="row">
            <h1 className="recommendationWizard__title small-12 column">
              What year were you born
            </h1>
            <div className="small-12 column">
              <div className="row">
                <div className="small-12 large-6 large-offset-3 column">
                  <Input
                    placeholder="e.g. 1980"
                    value={this.state.value}
                    type="number"
                    onChange={this.handleChange}
                  />
                </div>

                <div className="small-12 column">
                  <div className="recommendationWizard--center recommendationWizard__message">
                    {this.state.showMessage && 'you must enter a valid year'}
                  </div>
                </div>

                <div className="small-12 column">
                  <p className="recommendationWizard--center">
                    <a
                      href="/recommender/dumb-link"
                      className="recommendationWizard__link"
                      onClick={this.onOpenModal}
                    >
                      Why do we ask for year of birth?
                    </a>
                  </p>
                </div>

                <Button
                  className="recommendationWizard__button"
                  onClick={this.handleClick}
                >
                  Complete
                </Button>
              </div>
            </div>
          </div>
        </div>
        <BaseModal
          className="small-12 medium-6 column"
          isOpen={this.state.isModalOpen}
        >
          <h2>Why do we ask for year of birth?</h2>
          <p className="recommendationWizard__modalBody">
            Lift & Co. requests your age to help customize a cannabis product
            suggestion based on similar users within our database.
          </p>

          <div className="small-12 small-centered text-center columns">
            <Button onClick={this.onCloseModal}>Got it!</Button>
          </div>
        </BaseModal>
      </section>
    );
  }
}
