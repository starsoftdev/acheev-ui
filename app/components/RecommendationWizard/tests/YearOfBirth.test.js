import React from 'react';
import { create } from 'react-test-renderer';
import { shallow } from 'enzyme';
import YearOfBirth from 'components/RecommendationWizard/YearOfBirth';

describe('<YearOfBirth />', () => {
  it('should match snapshot', () => {
    const component = create(<YearOfBirth onSelect={() => {}} />);
    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('Should show error', () => {
    const handleSelect = jest.fn();
    const component = shallow(<YearOfBirth onSelect={handleSelect} />);

    component
      .find('Input')
      .first()
      .prop('onChange')('2018');

    component
      .find('Button')
      .first()
      .simulate('click');

    expect(component.state().showMessage).toEqual(true);
    expect(handleSelect).toHaveBeenCalledTimes(0);

    component
      .find('Input')
      .first()
      .prop('onChange')('1885');

    component
      .find('Button')
      .first()
      .simulate('click');

    expect(component.state().showMessage).toEqual(true);
    expect(handleSelect).toHaveBeenCalledTimes(0);
  });

  it('Should call onSelect', () => {
    const handleSelect = jest.fn();
    const component = shallow(<YearOfBirth onSelect={handleSelect} />);

    component
      .find('Input')
      .first()
      .prop('onChange')('1986');

    component
      .find('Button')
      .first()
      .simulate('click');

    expect(component.state().showMessage).toEqual(false);
    expect(handleSelect).toHaveBeenCalledWith('1986');
    expect(handleSelect).toHaveBeenCalledTimes(1);
  });
});
