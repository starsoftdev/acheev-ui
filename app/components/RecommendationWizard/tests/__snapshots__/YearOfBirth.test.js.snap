// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<YearOfBirth /> should match snapshot 1`] = `
<section
  className="recommendationWizard__body"
>
  <div
    className="container"
  >
    <div
      className="row"
    >
      <h1
        className="recommendationWizard__title small-12 column"
      >
        What year were you born
      </h1>
      <div
        className="small-12 column"
      >
        <div
          className="row"
        >
          <div
            className="small-12 large-6 large-offset-3 column"
          >
            <input
              onChange={[Function]}
              placeholder="e.g. 1980"
              type="number"
              value=""
            />
          </div>
          <div
            className="small-12 column"
          >
            <div
              className="recommendationWizard--center recommendationWizard__message"
            />
          </div>
          <div
            className="small-12 column"
          >
            <p
              className="recommendationWizard--center"
            >
              <a
                className="recommendationWizard__link"
                href="/recommender/dumb-link"
                onClick={[Function]}
              >
                Why do we ask for year of birth?
              </a>
            </p>
          </div>
          <a
            className="button recommendationWizard__button"
            onClick={[Function]}
          >
            <div
              className="row align-middle align-center"
            >
              <div
                className="shrink column"
              >
                Complete
              </div>
            </div>
          </a>
        </div>
      </div>
    </div>
  </div>
</section>
`;
